### PR TITLE
Display `SinceKotlin` everywhere

### PR DESCRIPTION
**Pull request by vmishenev**

### CODE REVIEW

**Review by IgnatBeresnev — COMMENTED** [review]
Since the PR is somewhat big, it would be easier and more productive if changes were broken into two different PRs: one for new `extraOptions` API and another one for `sinceKotlin`. There's little need to do it now, it's a request for the future

File: plugins/base/src/main/kotlin/transformers/documentables/ActualTypealiasAdder.kt [hunk 1]
```diff
@@ -67,12 +67,32 @@ class ActualTypealiasAdder : DocumentableTransformer {
...
-                        when(it) {
-                            is DClass -> it.copy(sourceSets = element.sourceSets + ta.sourceSets)
-                            is DEnum ->  it.copy(sourceSets = element.sourceSets + ta.sourceSets)
-                            is DInterface -> it.copy(sourceSets = element.sourceSets + ta.sourceSets)
-                            is DObject -> it.copy(sourceSets = element.sourceSets + ta.sourceSets)
-                            is DAnnotation -> it.copy(sourceSets = element.sourceSets + ta.sourceSets)
+                        when (it) {
+                            is DClass -> it.copy(
+                                documentation = element.documentation + ta.documentation,
+                                sourceSets = element.sourceSets + ta.sourceSets
```
[the remark this answers is not in the document]
Comment (vmishenev) [reply plugins/base/src/main/kotlin/transformers/documentables/ActualTypealiasAdder.kt:73]: Yes, The custom tag `SinceKotlin` keeps into `documentation`.

File: plugins/base/src/main/kotlin/transformers/pages/annotations/SinceKotlinTransformer.kt [hunk 2]
```diff
@@ -1,93 +1,167 @@
...
 import org.jetbrains.dokka.plugability.DokkaContext
 import org.jetbrains.dokka.transformers.documentation.DocumentableTransformer
+import org.jetbrains.dokka.utilities.associateWithNotNull
 import org.jetbrains.kotlin.utils.addToStdlib.safeAs
 
 class SinceKotlinTransformer(val context: DokkaContext) : DocumentableTransformer {
+    class Version constructor(str: String) : Comparable<Version> {
+        private val parts: List<Int> = str.split(".").map { it.toInt() }
+
+        override fun compareTo(other: Version): Int {
```
[the remark this answers is not in the document]
Comment (vmishenev) [reply plugins/base/src/main/kotlin/transformers/pages/annotations/SinceKotlinTransformer.kt:23]: See my unit test:
```
assert(Version("1.0.0").compareTo(Version("1")) == 0)
```

File: plugins/base/src/main/kotlin/transformers/pages/annotations/SinceKotlinTransformer.kt [hunk 3]
```diff
@@ -1,93 +1,167 @@
...
+                val diff = (if (i1.hasNext()) i1.next() else 0) - (if (i2.hasNext()) i2.next() else 0)
+                if (diff != 0) return diff
+            }
+
+            return 0
+        }
+
+        override fun toString(): String = parts.joinToString(".")
+    }
+    private val minSinceKotlin = mapOf(
```
[the remark this answers is not in the document]
Comment (vmishenev) [reply plugins/base/src/main/kotlin/transformers/pages/annotations/SinceKotlinTransformer.kt:37]: No, `SinceKotlin` does not mean a stable version.  
e.g. KMP is not stable.

File: runners/cli/src/main/kotlin/org/jetbrains/dokka/main.kt [hunk 4]
```diff
@@ -11,8 +11,12 @@ import java.nio.file.Paths
 class GlobalArguments(args: Array<String>) : DokkaConfiguration {
 
     val parser = ArgParser("globalArguments", prefixStyle = ArgParser.OptionPrefixStyle.JVM)
+    private val extraArgs: List<String> by parser.argument(ArgType.String, "Json file name or extra options", ).vararg().optional()
```
[the remark this answers is not in the document]
Comment (vmishenev) [reply runners/cli/src/main/kotlin/org/jetbrains/dokka/main.kt:14]: `extraOptions` have a view of CLI arguments. Using them directly can be more convenient than encoding them into options.

File: core/src/main/kotlin/configuration.kt [hunk 5]
```diff
@@ -269,4 +270,12 @@ fun ExternalDocumentationLink(
 ): ExternalDocumentationLinkImpl =
     ExternalDocumentationLink(url.let(::URL), packageListUrl?.let(::URL))
 
+interface Option
+interface FeatureOption : Option {
+    val name: String
+}
+
+fun DokkaConfiguration.containsOption(option: FeatureOption) =
+    extraOptions.firstOrNull { it.indexOf("-XX") == 0 && option.name == it.substring(3) } != null
```
[the remark this answers is not in the document]
Comment (vmishenev) [reply core/src/main/kotlin/configuration.kt:279]: - It does not limit users to implementing their own functions. Everyone can define an overload 
`fun DokkaConfiguration.containsOption(option: Option)`
- Besides, we must not encourage users to use the extra options instead of a standard configuration.
- >  With object, it's also kind of verbose:

Anyway, you can use your own short implementation:
```kt
    open class FeatureOptionImpl(override val name: String): FeatureOption
    object MyFeatureOption: FeatureOptionImpl("my-feature-opt")
```   
- > Necessary -XX start of the option is also a tight constraint - developers might have their own convention of option naming within the company, or would just like to do something different with meaning behind it.

See my PR comments above. 
- > Speaking of KGP, it looks like that's exactly the way it's implemented for `freeCompilerArgs`: no type safety, and consumers decide themselves how they want to check if a value is present.

KGP does not use arguments, it only calls a compiler with them.
Although, you can have a look under the hook of the compiler.  
https://github.com/JetBrains/kotlin/blob/master/compiler/cli/cli-common/src/org/jetbrains/kotlin/cli/common/arguments/InternalCompilerArgument.kt
https://github.com/JetBrains/kotlin/blob/master/compiler/util/src/org/jetbrains/kotlin/config/LanguageVersionSettings.kt

File: plugins/base/src/test/kotlin/content/annotations/SinceKotlinTest.kt [hunk 6]
```diff
@@ -18,6 +28,216 @@ class SinceKotlinTest : BaseAbstractTest() {
...
+        assert(Version("1.0") >= Version("1.0"))
+        assert(Version("1.1") > Version("1"))
+        assert(Version("1.0") < Version("2.0"))
+        assert(Version("1.0") < Version("2.2"))
+    }
+
+    @Test
+    fun `rendered SinceKotlin custom tag for typealias, extensions, functions, properties`() {
+        val configuration =   dokkaConfiguration {
+            extraOptions = listOf("-XXSinceKotlin")
```
[the remark this answers is not in the document]
Comment (IgnatBeresnev) [reply plugins/base/src/test/kotlin/content/annotations/SinceKotlinTest.kt:44]: From the Kotlin docs:

> The "-X" and "-XX" flags are experimental and can be added and removed at any time.

Since `-XX` is part of the [hardcoded parsing](https://github.com/Kotlin/dokka/pull/2708/files#diff-89910d7fe7909fca5a75f09a3e3d7b446eaf94e342b48f79a4e9c59147a7ef87R279) in current code, this would mean that all flags are experimental and can be added and removed at any time, which isn't user friendly and might certainly differ from the behaviour that external plugin writers want for their own flags.

I think it's fine to add it to `-XXSinceKotlin`, since the feature is specific to stdlib and is not for average users, but making all flags experimental by default with no way of making it stable can lead to misunderstandings, so I think it should be dropped.

File: plugins/base/src/main/kotlin/transformers/pages/annotations/SinceKotlinTransformer.kt (hunk 2 above)
[the remark this answers is not in the document]
Comment (IgnatBeresnev) [reply plugins/base/src/main/kotlin/transformers/pages/annotations/SinceKotlinTransformer.kt:22]: Can you please add a comment then, explaining why it was created and which corner cases it handles as it's not obvious at first sight

File: plugins/base/src/main/kotlin/transformers/pages/annotations/SinceKotlinTransformer.kt (hunk 3 above)
[the remark this answers is not in the document]
Comment (IgnatBeresnev) [reply plugins/base/src/main/kotlin/transformers/pages/annotations/SinceKotlinTransformer.kt:37]: So what is it then? min version of what?

File: plugins/base/src/main/kotlin/transformers/pages/annotations/SinceKotlinTransformer.kt [hunk 7]
```diff
@@ -1,93 +1,171 @@
...
             else -> this.also { context.logger.warn("Unrecognized documentable $this while SinceKotlin transformation") }
         }
+    }
+
+    private fun List<Annotations.Annotation>.findSinceKotlinAnnotation(): Annotations.Annotation? =
+        this.find { it.dri.packageName == "kotlin" && it.dri.classNames == "SinceKotlin" }
+
+    private fun Documentable.getVersion(sourceSet: DokkaConfiguration.DokkaSourceSet): Version {
+        val annotatedVersion =
+            safeAs<WithExtraProperties<Documentable>>()?.extra?.get(Annotations)?.directAnnotations?.get(sourceSet)
```
Comment (IgnatBeresnev) [inline plugins/base/src/main/kotlin/transformers/pages/annotations/SinceKotlinTransformer.kt:120]: As mentioned previously, I believe it can be substituted with an existing function `this.annotations()[sourceSet]`

File: plugins/base/src/main/kotlin/transformers/pages/annotations/SinceKotlinTransformer.kt [hunk 8]
```diff
@@ -1,93 +1,171 @@
...
+                ?.findSinceKotlinAnnotation()
+                ?.params?.get("version").safeAs<StringValue>()?.value
+                ?.dropWhile { it == '"' }?.dropLastWhile { it == '"' }
```
Comment (IgnatBeresnev) [inline plugins/base/src/main/kotlin/transformers/pages/annotations/SinceKotlinTransformer.kt:123]: Any idea what this is for? If this is dropped, do tests pass and does stdlib output look OK?

File: runners/cli/src/main/kotlin/org/jetbrains/dokka/main.kt (hunk 4 above)
[the remark this answers is not in the document]
Comment (IgnatBeresnev) [reply runners/cli/src/main/kotlin/org/jetbrains/dokka/main.kt:14]: But options and arguments are not the same thing. Consider the following use cases for when you want to add extra options "one", "two", "three" and "four":

`-java -jar dokka-cli.jar -sourceSetName name -offlineMode one two three four`

vs

`java -jar dokka-cli.jar -sourceSet name name -offlineMode -extraOptions one,two,three,four`

Looking at the first example, I cannot deduce that these are _options_ - they look like some input/otuput arguments to me. In the second one it's clear

Using one argument for two things, especially a vararg, can also be very error-prone and confusing, and closes the door on ever adding a second argument. 

For instance, if you mistype some basic Dokka option, it will be silently parsed as an extra option instead of saying that such option does not exist.

Or if the provided json path is not correct:

previously:
```
ignat@workstation ~ [1]> java -jar dokka-cli-1.7.20.jar filepath.json
Exception in thread "main" java.io.FileNotFoundException: filepath.json (No such file or directory)
        at java.base/java.io.FileInputStream.open0(Native Method)
        at java.base/java.io.FileInputStream.open(FileInputStream.java:219)
        at java.base/java.io.FileInputStream.<init>(FileInputStream.java:157)
        at kotlin.io.FilesKt__FileReadWriteKt.readText(FileReadWrite.kt:125)
        at kotlin.io.FilesKt__FileReadWriteKt.readText$default(FileReadWrite.kt:125)
        at org.jetbrains.dokka.MainKt.initializeConfiguration(main.kt:15)
        at org.jetbrains.dokka.MainKt.main(main.kt:9
```

after this change:
```
ignat@workstation ~ [1]> java -jar dokka-cli-1.8.0-extra-options-SNAPSHOT.jar filepath.json
PROGRESS: Initializing plugins
Loaded plugins: []
Loaded: [

]
Suppressed: [

]
PROGRESS: 
Exception in thread "main" java.lang.IllegalStateException: ExtensionPoint: org.jetbrains.dokka.CoreExtensions/generation was expected to have exactly one extension registered, but none was found.
        at org.jetbrains.dokka.plugability.DokkaContextConfigurationImpl$single$1.invoke(DokkaContext.kt:151)
        at org.jetbrains.dokka.plugability.DokkaContextConfigurationImpl.single(DokkaContext.kt:158)
        at org.jetbrains.dokka.DokkaGenerator$generate$1.invoke(DokkaGenerator.kt:27)
        at org.jetbrains.dokka.DokkaGenerator$generate$1.invoke(DokkaGenerator.kt:17)
        at org.jetbrains.dokka.DokkaGeneratorKt.timed(DokkaGenerator.kt:76)
        at org.jetbrains.dokka.DokkaGeneratorKt.access$timed(DokkaGenerator.kt:1)
        at org.jetbrains.dokka.DokkaGenerator.generate(DokkaGenerator.kt:22)
        at org.jetbrains.dokka.MainKt.main(main.kt:431)
```

File: core/src/main/kotlin/configuration.kt (hunk 5 above)
[the remark this answers is not in the document]
Comment (IgnatBeresnev) [reply core/src/main/kotlin/configuration.kt:279]: At the moment, `Option` / `FeatureOption` API is very limiting and specific. If the users have to implement everything themselves, I fail to see why it needs to be public. Especially given that `Option` is not used anywhere as type.

File: plugins/base/src/main/kotlin/transformers/pages/annotations/SinceKotlinTransformer.kt (hunk 2 above)
[the remark this answers is not in the document]
Comment (vmishenev) [reply plugins/base/src/main/kotlin/transformers/pages/annotations/SinceKotlinTransformer.kt:22]: > Can you please add a comment then, explaining why it was created and which corner cases it handles as it's not obvious at first sight

We have the unit test in this case.

> Also, does t his class need to leak into public API?

Yes, it is used in the version filter.

File: plugins/base/src/main/kotlin/transformers/pages/annotations/SinceKotlinTransformer.kt (hunk 3 above)
[the remark this answers is not in the document]
Comment (vmishenev) [reply plugins/base/src/main/kotlin/transformers/pages/annotations/SinceKotlinTransformer.kt:37]: I will rename it to `minSinceKotlinOfPlatform`

File: plugins/base/src/test/kotlin/content/annotations/SinceKotlinTest.kt (hunk 6 above)
[the remark this answers is not in the document]
Comment (vmishenev) [reply plugins/base/src/test/kotlin/content/annotations/SinceKotlinTest.kt:44]: It's better to change the name `FeatureOption` to `ExperimentalFeatureOption`. WDYT?

File: core/src/main/kotlin/configuration.kt (hunk 5 above)
[the remark this answers is not in the document]
Comment (vmishenev) [reply core/src/main/kotlin/configuration.kt:279]: Can you give me the user cases you expect from extra options? 
I agree that in some cases `key=value` can be needed, but it might be introduced later via expanding the current API.
In my opinion, all cases of using extra options should be relevant to experimental features that we do not use want to use in a public configuration of a plugin. In other cases, options from a plugin configuration should be used.

File: runners/cli/src/main/kotlin/org/jetbrains/dokka/main.kt (hunk 4 above)
[the remark this answers is not in the document]
Comment (vmishenev) [reply runners/cli/src/main/kotlin/org/jetbrains/dokka/main.kt:14]: > extra options "one", "two", "three" and "four":

They do not look like extra CLI options. In contrast, `-extraOptions -one,-two-,three,-four` can look weird without preprocessing. 

> For instance, if you mistype some basic Dokka option, it will be silently parsed as an extra option instead of saying that such option does not exist.

I agree It's a problem. Unfortunately, we do not check extra options at all.

> Using one argument for two things, especially a vararg, can also be very error-prone and confusing, and closes the door on ever adding a second argument.
> Or if the provided json path is not correct:

It is solvable.
 
 Anyway, we can discuss an option format and cases of using options.

File: core/src/main/kotlin/configuration.kt (hunk 5 above)
[the remark this answers is not in the document]
Comment (vmishenev) [reply core/src/main/kotlin/configuration.kt]: So the extra options are postponed until we have [more cases of using them](https://github.com/Kotlin/dokka/issues/2776).

File: plugins/base/src/main/kotlin/transformers/pages/annotations/SinceKotlinTransformer.kt (hunk 2 above)
[the remark this answers is not in the document]
Comment (IgnatBeresnev) [reply plugins/base/src/main/kotlin/transformers/pages/annotations/SinceKotlinTransformer.kt:22]: > We have the unit test in this case.

It leaks into Public API - unfortunately, our unit tests will not be visible from there

I would actually consider extracting it out of `SinceKotlinTransformer` into it's own class (e.g `SinceKotlinVersion`) to make it reusable by other plugins

File: plugins/base/src/main/kotlin/transformers/pages/annotations/SinceKotlinTransformer.kt [hunk 9]
```diff
@@ -1,93 +1,173 @@
...
+                if (diff != 0) return diff
+            }
+
+            return 0
+        }
+
+        override fun toString(): String = parts.joinToString(".")
+    }
+
+    private val minSinceKotlinOfPlatform = mapOf(
```
Comment (IgnatBeresnev) [inline plugins/base/src/main/kotlin/transformers/pages/annotations/SinceKotlinTransformer.kt:37]: `SinceKotlin` detail looks redundant in this case

```suggestion
    private val minVersionOfPlatform = mapOf(
```

File: plugins/base/src/main/kotlin/transformers/pages/annotations/SinceKotlinTransformer.kt [hunk 10]
```diff
@@ -1,93 +1,173 @@
...
-                                ),
-                                "Since Kotlin"
-                            )
+                            customTag
                         )
                     ) else it.value
                 }
-            } ?: acc
         }
+    companion object {
```
Comment (IgnatBeresnev) [inline plugins/base/src/main/kotlin/transformers/pages/annotations/SinceKotlinTransformer.kt:170]: Can the companion object be internal too?

File: plugins/base/src/main/kotlin/transformers/pages/annotations/SinceKotlinTransformer.kt [hunk 11]
```diff
@@ -1,93 +1,173 @@
...
+       internal fun isAddSinceKotlin() = System.getProperty("dokka.SinceKotlin") in listOf("true", "1")
```
Comment (IgnatBeresnev) [inline plugins/base/src/main/kotlin/transformers/pages/annotations/SinceKotlinTransformer.kt:171]: Without context and on its own, `isAddSinceKotlin` and `dokka.SinceKotlin` don't give a hint of what they're used for. I'd propose to add an action word to it - `should`, `enable`, `handle` or something similar. Then it would hint that it enables some feature

Case sensitivity looks a bit strange for the property - dokka is decapitalized, and `addSinceKotlin` is camel-case, which isn't used much in property files, it's usually all lower-case

```suggestion
       internal fun shouldHandleSinceKotlin() = System.getProperty("dokka.enable-since-kotlin") in listOf("true", "1")
```

File: plugins/base/src/test/kotlin/content/annotations/SinceKotlinTest.kt [hunk 12]
```diff
@@ -18,8 +28,208 @@ class SinceKotlinTest : BaseAbstractTest() {
...
     }
 
+    @BeforeEach
+    fun setSystemProperty() {
+        System.setProperty("dokka.SinceKotlin", "true")
+    }
+    @AfterEach
+    fun clearSystemProperty() {
+        System.setProperty("dokka.SinceKotlin", "")
+    }
```
Comment (IgnatBeresnev) [inline plugins/base/src/test/kotlin/content/annotations/SinceKotlinTest.kt:38]: Does it have to be set and unset for each test? Can it be substituted with `@BeforeAll` and `@AfterAll`?

File: plugins/base/src/test/kotlin/content/annotations/SinceKotlinTest.kt [hunk 13]
```diff
@@ -18,8 +28,208 @@ class SinceKotlinTest : BaseAbstractTest() {
...
+                        assertEquals((tag.children.first() as Text).body, i.value.toString())
+                    }
+                }
+            }
+        }
+    }
+
     @Test
     fun `function with since kotlin annotation`() {
+        System.setProperty("dokka.SinceKotlin", "")
```
Comment (IgnatBeresnev) [inline plugins/base/src/test/kotlin/content/annotations/SinceKotlinTest.kt:232]: Can't tell if it's intentional or a leftover - does this test check that there's no `SinceKotlin` handling if the property is not set? If so, can we reflect that in the name of the test?

File: plugins/base/src/main/kotlin/transformers/pages/annotations/SinceKotlinTransformer.kt [hunk 14]
```diff
@@ -1,93 +1,173 @@
...
+
+    private fun Documentable.appendSinceKotlin(versions: SourceSetDependent<Version>) =
         sourceSets.fold(documentation) { acc, sourceSet ->
-            safeAs<WithExtraProperties<Documentable>>()?.extra?.get(Annotations)?.directAnnotations?.get(sourceSet)?.find {
-                it.dri == DRI("kotlin", "SinceKotlin")
-            }?.params?.get("version").safeAs<StringValue>()?.value?.let { version ->
+
+            val version = versions[sourceSet]
+
+            val customTag = CustomTagWrapper(
```
Comment (IgnatBeresnev) [inline plugins/base/src/main/kotlin/transformers/pages/annotations/SinceKotlinTransformer.kt:148]:
```suggestion
            val customSinceKotlinTag = CustomTagWrapper(
```

File: plugins/base/src/main/kotlin/transformers/pages/annotations/SinceKotlinTransformer.kt (hunk 9 above)
Comment (vmishenev) [reply plugins/base/src/main/kotlin/transformers/pages/annotations/SinceKotlinTransformer.kt:37]: I would left it, naming costs nothing. `Version` can be confusing. A version of what? Library!? Dokka!? Kotlin?!

File: plugins/base/src/main/kotlin/transformers/pages/annotations/SinceKotlinTransformer.kt (hunk 11 above)
Comment (vmishenev) [reply plugins/base/src/main/kotlin/transformers/pages/annotations/SinceKotlinTransformer.kt:171]: Camel case for properties stems from Gradle:
```
systemProp.gradle.wrapperUser=myuser
systemProp.gradle.wrapperPassword=mypassword
```

Comment (vmishenev) [reply plugins/base/src/main/kotlin/transformers/pages/annotations/SinceKotlinTransformer.kt:171]: Anyway, I will rename it to kebab-case.

File: plugins/base/src/test/kotlin/content/annotations/SinceKotlinTest.kt (hunk 12 above)
Comment (vmishenev) [reply plugins/base/src/test/kotlin/content/annotations/SinceKotlinTest.kt:38]: Some tests can unset the property.

File: plugins/base/src/main/kotlin/transformers/pages/annotations/SinceKotlinTransformer.kt (hunk 9 above)
Comment (IgnatBeresnev) [reply plugins/base/src/main/kotlin/transformers/pages/annotations/SinceKotlinTransformer.kt]: > A version of a platform?

It's a **min** version of a platform, yeah. As in the "first version in which this platform appeared" - as far as I understand, that's what this property is used for. `common` first appeared in 1.2, `native` was announced in 1.3, `js` in 1.1 (I looked through the "What's changed" archive)

The reason I suggested to remove `SinceKotlin` from it is because it makes you think if `SinceKotlin` is an exceptional case or it has some meaning, like maybe native target was actually introduced in Kotlin 1.1, but min version of `Since Kotlin` of native target is for some reason 1.3. Whereas, as I understand it, `SinceKotlin` adds no explanation to the used version

So the less confusing words that make you question things, the better, imo

File: plugins/base/src/main/kotlin/transformers/pages/annotations/SinceKotlinTransformer.kt [hunk 15]
```diff
@@ -1,93 +1,177 @@
...
-import org.jetbrains.dokka.model.properties.WithExtraProperties
 import org.jetbrains.dokka.plugability.DokkaContext
 import org.jetbrains.dokka.transformers.documentation.DocumentableTransformer
+import org.jetbrains.dokka.utilities.associateWithNotNull
 import org.jetbrains.kotlin.utils.addToStdlib.safeAs
 
+class SinceKotlinVersion constructor(str: String) : Comparable<SinceKotlinVersion> {
+    private val parts: List<Int> = str.split(".").map { it.toInt() }
+
+    override fun compareTo(other: SinceKotlinVersion): Int {
```
Comment (IgnatBeresnev) [inline plugins/base/src/main/kotlin/transformers/pages/annotations/SinceKotlinTransformer.kt:21]: I still think it's worth documenting :)

**Review by IgnatBeresnev — APPROVED** [review]
